src/utils/yupUtils.ts

```diff
@@ -0,0 +1,74 @@
+import * as Yup from "yup";
+import { MixedLocale } from "yup/lib/locale";
+
+interface YupCustomSchema<In, C, Out = In> extends Yup.BaseSchema<In, C, Out> {
+  required(
+    msg?: MixedLocale["required"]
+  ): YupCustomSchema<In, C, NonNullable<Out>>;
+  nullable(isNullable?: true): YupCustomSchema<In | null, C>;
+  nullable(isNullable: false): YupCustomSchema<Exclude<In, null>, C>;
+}
+
+export function YupLiteral<
+  Literal extends string | number | boolean | null | undefined
+>(literal: Literal): YupCustomSchema<Literal, any, Literal> {
+  return Yup.mixed().oneOf([literal]);
+}
+
+export function YupUnion<Options extends Array<Yup.BaseSchema>>(
+  options: Options
+): YupCustomSchema<
+  Yup.TypeOf<Options[number]>,
+  any,
+  Yup.InferType<Options[number]>
+> {
+  return Yup.mixed().test(
+    "union",
+    "Must match one of the provided schemas",
+    function (obj) {
+      const { path, createError } = this;
+      for (const option of options) {
+        try {
+          option.validateSync(obj);
+          return true;
+        } catch (err) {
+          // Ignore validation error
+        }
+      }
+      return createError({
+        path,
+        message: "Must match one of the provided schemas"
+      });
+    }
+  ) as any;
+}
```

Comment:
You're using `validateSync`, which doesn’t handle async schemas. 
Optional async version
```suggestion
export async function YupUnionAsync<Options extends Array<Yup.BaseSchema>>(
  options: Options
): Promise<Yup.BaseSchema> {
  return Yup.mixed().test(
    "union",
    "Must match one of the provided schemas",
    async function (obj) {
      const { path, createError } = this;
      for (const option of options) {
        try {
          await option.validate(obj);
          return true;
        } catch {
          // Continue
        }
      }
      return createError({
        path,
        message: "Must match one of the provided schemas"
      });
    }
  );
}

```

The comment `// Ignore validation error` does not explain why the error is ignored.

Reply:
related 2f20217c8e63370902ede3353b57aba3d2734018